src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+    AkIdNum,
+    AkId,
+    TInfo,
+    KeySize,
+    Quote,
+}
+
+impl ReqType {
+    fn set_request(
+        &self,
+        report: Option<&[u8; 432]>,
+        akid: Option<Vec<u8>>,
+        size: Option<usize>,
+    ) -> Result<Request, Error> {
+        let mut req = Request::new();
+
+        match self {
+            ReqType::AkIdNum => {
+                // removed
+                unimplemented!()
+            }
+            ReqType::AkId => {
+                // removed
+                unimplemented!()
+            }
+            ReqType::TInfo => {
+                let akid = akid.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing attestation key ID from InitQuoteExRequest",
+                    )
+                })?;
+                let size = size.ok_or_else(|| {
+                    Error::new(ErrorKind::Other, "Missing key size from InitQuoteExRequest")
+                })?;
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(AESM_REQUEST_TIMEOUT);
+                msg.set_b_pub_key_id(true);
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(size as u64);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::KeySize => {
+                let akid = akid.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing attestation key ID from InitQuoteExRequest",
+                    )
+                })?;
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(AESM_REQUEST_TIMEOUT);
+                msg.set_b_pub_key_id(false);
+                msg.set_att_key_id(akid);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::Quote => {
+                let akid = akid.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing attestation key ID from InitQuoteExRequest",
+                    )
+                })?;
+                let report = report.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing REPORT structure from GetQuoteExRequest",
+                    )
+                })?;
+                let mut msg = Request_GetQuoteExRequest::new();
+                msg.set_timeout(AESM_REQUEST_TIMEOUT);
+                msg.set_report(report.to_vec());
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(SGX_QUOTE_SIZE as u32);
+                req.set_getQuoteExReq(msg);
+                Ok(req)
+            }
+        }
+    }
+
+    fn send_request(&self, req: Request, stream: &mut UnixStream) -> Result<Response, Error> {
+        // Set up writer
+        let mut buf_wrtr = vec![0u8; size_of::<u32>()];
```

Comment:
It is implemented for a mutable borrow of a slice https://doc.rust-lang.org/std/primitive.slice.html#impl-Write
And https://doc.rust-lang.org/std/primitive.array.html#method.as_mut on an array returns a `&mut [u8]`, therefore `buf_wrtr.as_mut()` implements a `Writer`, where `let buf_wrtr = [0u8; 4]`

Also, `_wrtr` suffix only makes it less clear what it is, `buf` would be way more conventional here

Reply:
```
        // Set up writer
        let mut buf_wrtr = [0u8; 4].as_mut();

        req.write_to_writer(buf_wrtr).map_err(|e| Error::new(
            ErrorKind::Other,
            format!("Invalid protobuf request: {:?}. Error: {:?}", req, e),
        ))?;
```

Result:

```
   Compiling enarx v0.2.0 (/home/jarkko/Projects/Enarx/enarx)
error[E0277]: the trait bound `[u8]: std::io::Write` is not satisfied
   --> src/backend/sgx/attestation.rs:112:29
    |
112 |         req.write_to_writer(buf_wrtr).map_err(|e| Error::new(
    |                             ^^^^^^^^ the trait `std::io::Write` is not implemented for `[u8]`
    |
    = help: the following implementations were found:
              <&mut [u8] as std::io::Write>
    = note: required for the cast to the object type `dyn std::io::Write`

error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
   --> src/backend/sgx/attestation.rs:112:29
    |
112 |         req.write_to_writer(buf_wrtr).map_err(|e| Error::new(
    |                             ^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[u8]`
    = note: required for the cast to the object type `dyn std::io::Write`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `enarx` due to 2 previous errors
```

Reply:
Alright, so what does this actually do?

First, we write the raw protobuf bytes into the buffer. We don't know the size of the message we're going to read, therefore we *do* need the vector. https://github.com/enarx/enarx/blob/81fbf186cb33c08054352b74a025d0ebbb9bf0ef/src/protobuf/aesm-proto.proto#L37-L138

After that, on L59 we overwrite the first 4 bytes of the buffer with length of the request.

So I guess the intention is to store the length of the request in the first 4 bytes and append the actual protobuf message to the vector?

Why do we need this, if we already have a protobuf? Is this some specific binary format?
This most definitely lacks some documentation.

Reply:
Looks like we should use https://docs.rs/protobuf/latest/protobuf/trait.Message.html#method.write_length_delimited_to_vec here instead